### PR TITLE
[Discover] Add long numerals support

### PR DESCRIPTION
**Pull request by AMoo-Miki**
### Description

* Add long numerals support to Discover tabular view, document view, inspector, hash-url, and share
* Use a fork of `numeral-js` that supports long numerals
* Patch `rison` to correctly serialize BigInts
Replace the uses of `withLongNumerals` with `withLongNumeralsSupport` to match the rest of the codebase and mark `withLongNumerals` for deprecation
* Add `withLongNumeralsSupport` as an option to `code/public/http/fetch`
* Add long numerals support to `UiSettingsClient`
* Add long numerals support to `core/server/http/router` response handler
* Add long numerals support to `RangeFilter` and `FilterBar`
* Add long numerals support to `kuery/ast`
* Introduce `OPENSEARCH_SEARCH_WITH_LONG_NUMERALS_STRATEGY` in `core/plugins/data/search`

### Check List

- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.tsx [hunk 1]
```diff
@@ -58,12 +59,12 @@ export const fetchTableDataCell = (
     return <span>-</span>;
   }
 
-  if (!fieldInfo?.type && flattenedRow && typeof flattenedRow[columnId] === 'object') {
+  if (!fieldInfo?.type && typeof flattenedRow?.[columnId] === 'object') {
```
Comment (kavilla) [inline src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.tsx:62]: does javascript end up equating this as undefined and null but then typeof of null equating to an object?

Comment (AMoo-Miki) [reply src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.tsx:62]: `flattenedRow?.[columnId]` would be `undefined` if `flattenedRow` was `null`, `undefined`, `Boolean(false)`, `Number(0)`, or `""`. For the last 3, the prototype props and functions can be named in `columnId` but then, they could be called with any `Boolean`, `Number`, or `String` and if we wanted to block those from being used, we would need to handle them for all of those types and not just for the falsy ones.

File: src/plugins/data/common/field_formats/converters/numeral.ts [hunk 2]
```diff
@@ -56,11 +56,14 @@ export abstract class NumeralFormat extends FieldFormat {
   protected getConvertedValue(val: any): string {
     if (val === -Infinity) return '-∞';
     if (val === +Infinity) return '+∞';
-    if (typeof val !== 'number') {
+
+    const isBigInt = typeof val === 'bigint';
```
Comment (ruanyl) [inline src/plugins/data/common/field_formats/converters/numeral.ts:60]: Does it make sense to add a unit test for bigint?

File: package.json [hunk 3]
```diff
@@ -129,7 +129,7 @@
     "@elastic/datemath": "5.0.3",
     "@elastic/eui": "npm:@opensearch-project/oui@1.4.0-alpha.2",
     "@elastic/good": "^9.0.1-kibana3",
-    "@elastic/numeral": "^2.5.0",
+    "@elastic/numeral": "npm:@amoo-miki/numeral@2.6.0",
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply package.json:132]: Yulong, it is temporary for OSD 2.x and I am hoping to switch to `numerable` in 3.x which hasn't seen change in 3 years but the maintainers almost actively reply to issues. My hope is that I will add `BigInt` support to `numerable` and then we can just depend on it; if not, we will just fold it in to `packages`.

File: scripts/postinstall.js [hunk 4]
```diff
@@ -71,6 +71,19 @@ const run = async () => {
       },
     ])
   );
+  promises.push(
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply scripts/postinstall.js:74]: Kinda. I decided to go with this super primitive approach until OSD 3.x which will use `yarn@2` aka `berry` which has built-in patching capabilities.

File: src/plugins/data/common/opensearch_query/filters/range_filter.ts [hunk 5]
```diff
@@ -124,19 +124,24 @@ export const buildRangeFilter = (
     filter.meta.formattedValue = formattedValue;
   }
 
-  params = mapValues(params, (value: any) => (field.type === 'number' ? parseFloat(value) : value));
+  params = mapValues(params, (value: any) =>
+    field.type === 'number'
+      ? isFinite(value) && (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER)
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply src/plugins/data/common/opensearch_query/filters/range_filter.ts:129]: While there is no entry way for BigInts to reach this, I like the idea of guarding against it; I will add that.

File: src/plugins/data/common/opensearch_query/filters/range_filter.ts [hunk 6]
```diff
@@ -124,19 +124,24 @@ export const buildRangeFilter = (
     filter.meta.formattedValue = formattedValue;
   }
 
-  params = mapValues(params, (value: any) => (field.type === 'number' ? parseFloat(value) : value));
+  params = mapValues(params, (value: any) =>
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply src/plugins/data/common/opensearch_query/filters/range_filter.ts:127]: On their way.

File: src/plugins/data/common/field_formats/converters/numeral.ts (hunk 2 above)
Comment (AMoo-Miki) [reply src/plugins/data/common/field_formats/converters/numeral.ts:60]: On their way.

File: scripts/postinstall.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruanyl) [reply scripts/postinstall.js:74]: I see, thanks for the clarification 👍

File: src/core/public/http/types.ts [hunk 7]
```diff
@@ -262,7 +262,7 @@ export interface HttpFetchOptions extends HttpRequestInit {
    * When `true`, if the response has a JSON mime type, the {@link HttpResponse} will use an alternate JSON parser
    * that converts long numerals to BigInts. Defaults to `false`.
    */
-  withLongNumerals?: boolean;
+  withLongNumeralsSupport?: boolean;
```
Comment (BSFishy) [inline src/core/public/http/types.ts:265]: If I'm not mistaken, this is part of core's public API. Are we concerned with this being breaking?

File: src/core/public/ui_settings/ui_settings_client.ts [hunk 8]
```diff
@@ -97,11 +97,11 @@ You can use \`IUiSettingsClient.get("${key}", defaultValue)\`, which will just r
...
-    if (type === 'number') {
-      return parseFloat(value);
-    }
-
-    return value;
+    return type === 'number' && typeof value !== 'bigint'
+      ? isFinite(value) && (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER)
+        ? BigInt(value)
+        : parseFloat(value)
+      : value;
```
Comment (BSFishy) [inline src/core/public/ui_settings/ui_settings_client.ts:104]: nit: I prefer if-else over complex ternaries for readability

File: src/plugins/data/common/opensearch_query/filters/range_filter.ts [hunk 9]
```diff
@@ -124,19 +124,24 @@ export const buildRangeFilter = (
...
+    field.type === 'number' && typeof value !== 'bigint'
+      ? isFinite(value) && (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER)
+        ? BigInt(value)
+        : parseFloat(value)
+      : value
```
Comment (BSFishy) [inline src/plugins/data/common/opensearch_query/filters/range_filter.ts:132]: This type of ternary seems to pop up a few times. Have you considered pulling it out into a util function?

File: src/core/public/http/types.ts (hunk 7 above)
Comment (AMoo-Miki) [reply src/core/public/http/types.ts:265]: Yeah; i did this when 2.11 was not released. I updated the code and marked it deprecated. I reintroduced the consumption of it and added it to the deprecation changelog.

File: src/plugins/data/common/opensearch_query/filters/range_filter.ts (hunk 9 above)
Comment (AMoo-Miki) [reply src/plugins/data/common/opensearch_query/filters/range_filter.ts:132]: Yeah. One was in core and one was in a plugin so I let it slide. If I see it again, I will pull it into a util.

**Review by BSFishy — APPROVED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -21,6 +23,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Workspace] Add core workspace service module to enable the implementation of workspace features within OSD plugins ([#5092](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5092))
 - [Workspace] Setup workspace skeleton and implement basic CRUD API ([#5075](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5075))
 - [Decouple] Add new cross compatibility check core service which export functionality for plugins to verify if their OpenSearch plugin counterpart is installed on the cluster or has incompatible version to configure the plugin behavior([#4710](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4710))
+- [Discover] Add long numerals support [#6829](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6829)
```
Comment (manasvinibs) [inline CHANGELOG.md:26]: nit: PR number

Comment (AMoo-Miki) [reply CHANGELOG.md:26]: How on earth did that happen!?

File: CHANGELOG.md [hunk 11]
```diff
@@ -8,6 +8,8 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ### Deprecations
 
+- Rename `withLongNumerals` to `withLongNumeralsSupport` in `HttpFetchOptions` [#6829](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6829)
```
Comment (manasvinibs) [inline CHANGELOG.md:11]: what about this one?

Comment (AMoo-Miki) [reply CHANGELOG.md:11]: :/ that too. Thanks.

**Review by kavilla — APPROVED** [review]

**Review by manasvinibs — APPROVED** [review]

**Pull request merged**